mm2src/mm2_main/src/lp_healthcheck.rs

```diff
@@ -0,0 +1,505 @@
+        // That's us, so return true.
+        return Ok(true);
+    }
+
+    let message = HealthcheckMessage::generate_message(
+        &ctx,
+        target_peer_address,
+        false,
+        ctx.health_checker
+            .config
+            .message_expiration_secs
+            .try_into()
+            .map_err(|e: TryFromIntError| HealthcheckRpcError::Internal { reason: e.to_string() })?,
+    )
+    .map_err(|reason| HealthcheckRpcError::MessageGenerationFailed { reason })?;
+
+    let encoded_message = message
+        .encode()
+        .map_err(|e| HealthcheckRpcError::MessageEncodingFailed { reason: e.to_string() })?;
+
+    let (tx, rx): (Sender<()>, Receiver<()>) = oneshot::channel();
+
+    {
+        let mut book = ctx.health_checker.response_handler.lock().await;
+        book.clear_expired_entries();
+        book.insert(target_peer_address.to_string(), tx, address_record_exp);
+    }
+
+    broadcast_p2p_msg(
+        &ctx,
+        peer_healthcheck_topic(&target_peer_address),
+        encoded_message,
+        None,
+    );
+
+    let timeout_duration = Duration::from_secs(ctx.health_checker.config.timeout_secs);
+    Ok(rx.timeout(timeout_duration).await == Ok(Ok(())))
+}
+
+pub(crate) async fn process_p2p_healthcheck_message(ctx: &MmArc, message: mm2_libp2p::GossipsubMessage) {
+    macro_rules! try_or_return {
+        ($exp:expr, $msg: expr) => {
+            match $exp {
+                Ok(t) => t,
+                Err(e) => {
+                    log::error!("{}, error: {e:?}", $msg);
+                    return;
+                },
+            }
+        };
+    }
+
+    let data = try_or_return!(
+        HealthcheckMessage::decode(&message.data),
+        "Couldn't decode healthcheck message"
+    );
+
+    let sender_peer = data.sender_peer();
+
+    let ctx = ctx.clone();
+
+    // Pass the remaining work to another thread to free up this one as soon as possible,
+    // so KDF can handle a high amount of healthcheck messages more efficiently.
+    ctx.spawner().spawn(async move {
+        let my_peer_address = P2PContext::fetch_from_mm_arc(&ctx).peer_id().into();
+
+        if !data.is_received_message_valid(my_peer_address, &ctx.health_checker.config) {
+            log::error!("Received an invalid healthcheck message.");
+            log::debug!("Message context: {:?}", data);
+            return;
+        };
+
+        let mut ddos_shield = ctx.health_checker.ddos_shield.lock().await;
+        ddos_shield.clear_expired_entries();
+        if ddos_shield
+            .insert(
+                sender_peer.to_string(),
+                (),
+                Duration::from_millis(ctx.health_checker.config.blocking_ms_for_per_address),
+            )
+            .is_some()
+        {
+            log::warn!("Peer '{sender_peer}' exceeded the healthcheck blocking time, skipping their message.");
+            return;
+        }
+        drop(ddos_shield);
```

Comment:
One last mumble from my side, and it's nit (don't wanna hold the PR more & we can add it later if we want).

This ddos shield could be bypassed if an attacker wants to. They can use many different sender peer keys and still DDoS.

what about having a general response that doesn't depend on who is the sender/requester. it only commits to being connected to a specific network (so the response mentions the network specification [netid?] and signs that). Same response could be reused for different requests from the same or different requesting peers. The same response could be used till it's expired then a new one should be generated on the next HC request.

Reply:
> Ideally we could limit receiving the maximum amount of request in specific time frames (e.g., 5requests/s, 50requests/60s, etc.) but that's entirely different scope of work.

This has false-positive results, like I mentioned it's very hard problem to solve by its nature.

Reply:
in our context, signing is an expensive op, and it's directly triggered from an outsider party so this is a space for ddosing. The shield is helpful but could go so far.

what i meant above was instead of having to preform an expensive op for each request, we only have to perform an expensive op periodically every expiration time `t`. this eliminates the DDoSing door here.

Reply:
Yeah we could reuse the message if expiration time is not too close. But still, this doesn't solve the problem completely as expiration times can be very short depending on the configuration. The protection implementation should be implemented directly at network level.

Reply:
Lmk if [this](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2194/commits/d2d9fff4685ce4f5edb41fbe96d61a0a667844f9) is acceptable.

Reply:
noooo. I meant a single message that could be used as a reply to any peer proving that we are connected to network `N`. for such a thing we need to remove the spetialization of target peer from such a message to be general and accepted by any peer as a proof of health in such network.

Reply:
noooo. I meant a single message that could be used as a reply to any peer proving that we are connected to network `N`. for such a thing we need to remove the spetialization of target peer from such a message to be general and accepted by any peer as a proof of health in such network.

Reply:
`target_peer` is intentionally included to make sure that signed message is uniquely generated for our request and not reused elsewhere. But if we remove the configurable expiration time and set it to a fixed value, we can eliminate `target_peer` from the signed message and make itmore consistent across all peers in the network. I will try it.

Reply:
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2194/commits/b98832c2eb46360588117555a4330b90ae69dfcb should make it cheaper and cachable. Config field is also removed to make it self-contained.